system-helper/flatpak-system-helper.c

```diff
@@ -217,6 +306,28 @@ flatpak_invocation_return_error (GDBusMethodInvocation *invocation,
     }
 }
 
+static OngoingPull *
+take_ongoing_pull_by_dir (const gchar *src_dir)
+{
+  OngoingPull *pull = NULL;
+  gpointer key, value;
+
+  G_LOCK (cache_dirs_in_use);
+  /* Keep src_dir key inside hashtable but remove its OngoingPull
+   * value and set it to NULL. This way src_dir is still marked
+   * as in-use (as Deploy or CancelPull might be executing on it,
+   * whereas OngoingPull ownership is transferred to respective
+   * callers. */
+  if (g_hash_table_steal_extended (cache_dirs_in_use, src_dir, &key, &value))
+    {
+      g_hash_table_insert (cache_dirs_in_use, key, NULL);
```

Comment:
Not that it affects anything, but it might be more clear if you add an if (value) g_hash_table_insert(..., NULL);